rtrlib/aspa/aspa.c

```diff
@@ -0,0 +1,267 @@
+/*
+ * This file is part of RTRlib.
+ *
+ * This file is subject to the terms and conditions of the MIT license.
+ * See the file LICENSE in the top level directory for more details.
+ *
+ * Website: http://rtrlib.realmv6.org/
+ */
+
+#include "aspa_private.h"
+#include "ordered_dyn_array/ordered_dyn_array.h"
+#include "rtrlib/rtrlib_export_private.h"
+#include "rtrlib/lib/alloc_utils_private.h"
+
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+
+static void aspa_table_notify_clients(struct aspa_table *aspa_table, const struct aspa_record *record, const struct rtr_socket *rtr_socket, const bool added)
+{
+	if (aspa_table->update_fp) {
+		// Realloc in order not to expose internal record
+		struct aspa_record *rec = lrtr_realloc((void *)record, sizeof(record) + record->provider_count * sizeof(record->provider_asns[0]));
+		aspa_table->update_fp(aspa_table, *rec, rtr_socket, added);
+	}
+}
+
+RTRLIB_EXPORT void aspa_table_init(struct aspa_table *aspa_table, aspa_update_fp update_fp)
+{
+	aspa_table->update_fp = update_fp;
+	pthread_rwlock_init(&(aspa_table->lock), NULL);
+}
+
+int aspa_table_cache_init(struct aspa_table *aspa_table, struct rtr_socket *rtr_socket)
+{
+	// ASPA table with cache does not need update_fp
+	aspa_table->update_fp = NULL;
+	
+	// Create an aspa_cache
+	// Records associated to the given rtr_socket won't be added or removed from the
+	// socket's ordered_dyn_array, but from the cache instead.
+	struct aspa_cache *cache = lrtr_malloc(sizeof(struct aspa_cache));
+	
+	// Save a reference to the socket, since the aspa cache table won't allow
+	// adding or removing records associated to other sockets!
+	cache->rtr_socket = rtr_socket;
+	
+	// copy the socket's existing ordered_dyn_array into the cache
+	if (ordered_dyn_array_copy(&(cache->cached_array), rtr_socket->aspa_array) < 0)
```

Comment:
If dont really like it when the `{}` are omitted

Reply:
I don't either but these kernel coding guidelines state https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces

Reply:
I can change that, what do you think @mrzslz? Concerns all of these guidelines

Reply:
the guidelines say: no curly braces here, and the contribution guidelines require this style, so no room for discussion here